src/azul/terra.py

```diff
@@ -353,8 +461,76 @@ def _repository_endpoint(self, *path: str) -> str:
         return furl(config.tdr_service_url,
                     path=('api', 'repository', 'v1', *path)).url
 
+    def _check_response(self,
+                        endpoint: str,
+                        headers: JSON = None,
+                        params: JSON = None
+                        ) -> MutableJSON:
+        response = self._request('GET', endpoint, headers=headers, fields=params)
+        if response.status == 200:
+            return json.loads(response.data)
+        elif response.status == 401:
+            self._on_insufficient_access(endpoint)
+        else:
+            raise RequirementError('Unexpected response from TDR API', response.status)
+
+    def list_snapshots(self) -> JSONs:
+        auth_key = self.auth_provider.auth_key()
+        table_name = config.dynamo_tdr_user_snapshots_table_name
+        response = self._dynamodb.get_item(TableName=table_name,
+                                           Key={self.ddb_key_name: {'S': auth_key}},
+                                           ProjectionExpression=','.join([self.ddb_value_name, self.ddb_ttl_name]))
+        result = response.get('Item')
+        now = int(time())
+        if result is None:
+            log.info('Snapshot list cache miss, %r', auth_key)
+        # Items can persist in DynamoDB after they are marked as expired
+        # https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/howitworks-ttl.html
+        elif int(result[self.ddb_ttl_name]['N']) < now:
+            log.info('Snapshot list cache is expired, %r', auth_key)
+            result = None
+        if result is None:
+            result = self._list_snapshots()
+            self._dynamodb.put_item(TableName=table_name,
+                                    Item={
+                                        self.ddb_key_name: {'S': auth_key},
+                                        self.ddb_value_name: {'S': json.dumps(result)},
+                                        self.ddb_ttl_name: {
+                                            'N': str(int(time()) + self.auth_provider.auth_key_lifespan())
```

Comment:
This TTL might be too long, if we cache the snapshots and a user gets access right after we perform the caching, they have to wait for the the `auth_key_lifespan` in order to see the new snapshots they now have access to.

Might want to discuss this w/ @hannes-ucsc

Reply:
I'll bring it up at parking lot unless @hannes-ucsc replies here.

Reply:
I decided not to bring this up at paring because I realized an obvious solution: since the snapshots are cached by access token and not by user email, users can reset/bypasss the cache simply by re-authenticating to obtain a fresh token. I think "logging out and logging back in again" is a sufficiently obvious course of action in the event of an authorization update that we don't need to change the implementation here.

Reply:
It an interesting UX case.. I would like to see how we can tackle this, might have to do some combination of reduced TTL and a fallback to  `logging out and logging back in again`.

How can we make it apparent that a client should `logging out and logging back in again`? I would think by the time they email support, and support pings the team, the cache could be invalidated. 

On a side note, maybe the new `/repository/sources` endpoint can take a `refresh` url parameter. Presence of this parameter would return the accessible sources from the repository rather than from our cache, and update the cache in the process.

Reply:
@hannes-ucsc: "Add FIXME to streamline cache line invalidation. Two events should invalidate the cache line: addition or removal of a snapshot, and a change in the set of snapshots that the user has access to."